{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import logging\n",
    "import re\n",
    "import pathlib\n",
    "\n",
    "import datetime\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from pyplink import PyPlink\n",
    "\n",
    "from basic_tools import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_KCHIP_HLA_AA_SNP_1000G=PyPlink(plink_KCHIP_HLA_AA_SNP_1000G_path)\n",
    "plink_KCHIP_HLA_AA_SNP_1000G_fam=plink_KCHIP_HLA_AA_SNP_1000G.get_fam().astype({'fid':str,'iid':str}).rename(columns={'fid':'FID','iid':'IID'})\n",
    "plink_KCHIP_HLA_AA_SNP_1000G_bim=plink_KCHIP_HLA_AA_SNP_1000G.get_bim()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_bed_path='data/known_genes_chr6.hg18.txt'\n",
    "gene_bed=pd.read_csv(gene_bed_path,sep='\\t')\n",
    "gene_bed=gene_bed[(gene_bed['txStart']>=plink_KCHIP_HLA_AA_SNP_1000G_bim.pos.min())&(gene_bed['txEnd']<=plink_KCHIP_HLA_AA_SNP_1000G_bim.pos.max())]\n",
    "\n",
    "gene_assign=plink_KCHIP_HLA_AA_SNP_1000G_bim[['pos']]\n",
    "\n",
    "for idx,row in gene_bed.iterrows():\n",
    "    gene_assign[row['name2']]=0\n",
    "    \n",
    "for idx,row in gene_bed.iterrows():    \n",
    "    gene_assign[row['name2']][(gene_assign['pos']>=row['txStart'])&(gene_assign['pos']<=row['txEnd'])]=1\n",
    "    \n",
    "HLA_names=np.unique([i[0].split('_')[1] for i in plink_KCHIP_HLA_AA_SNP_1000G_bim[plink_KCHIP_HLA_AA_SNP_1000G_bim.index.str.contains('HLA_')].index.str.split('*')])\n",
    "\n",
    "for HLA_name in HLA_names:\n",
    "    #print(HLA_name)\n",
    "    HLA_name='HLA-{}'.format(HLA_name)\n",
    "    #gene_assign_HLA[HLA_name]=0\n",
    "    gene_select=gene_assign[gene_assign.index.str.contains('HLA_'+HLA_name)|gene_assign.index.str.contains('SNPS_'+HLA_name)|gene_assign.index.str.contains('AA_'+HLA_name)]#print(gene_select.sort_values('pos').iloc[0],gene_select.sort_values('pos').iloc[-1])\n",
    "    gene_assign[HLA_name][(gene_assign['pos']>=gene_select['pos'].min())&(gene_assign['pos']<=gene_select['pos'].max())]=1\n",
    "    #txStart=gene_bed[gene_bed['name2']=='HLA-{}'.format(HLA_name)]['txStart']\n",
    "    ##txEnd=gene_bed[gene_bed['name2']=='HLA-{}'.format(HLA_name)]['txEnd']\n",
    "    #print(int(txStart),gene_assign_HLA[gene_assign_HLA[HLA_name]==1]['pos'].min(),gene_assign_HLA[gene_assign_HLA[HLA_name]==1]['pos'].max(),int(txEnd))     \n",
    "\"\"\"\n",
    "for idx,row in gene_bed.iterrows():\n",
    "    exon_range_list=list(zip(row['exonStarts'].strip(',').split(','),row['exonEnds'].strip(',').split(',')))\n",
    "    for start,end in exon_range_list:\n",
    "        start=int(start);end=int(end)\n",
    "        gene_assign_nonHLA_bool=((gene_assign_nonHLA['pos']>=start) & (gene_assign_nonHLA['pos']<=end))\n",
    "        gene_assign_nonHLA[row['name2']][gene_assign_nonHLA_bool]=1        \n",
    "\"\"\"            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes=pd.read_csv(pheno_all_file_path,sep='\\t').set_index('ID')\n",
    "binary_continuous_traits=sorted(phenotypes.columns[~phenotypes.columns.str.contains('x_ray')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALP continuous\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ch6845/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/ipykernel_launcher.py:17: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for phenotype_name in binary_continuous_traits:\n",
    "    \n",
    "    pheno=pd.read_csv(data_out_pheno_path+phenotype_name+'.phe',sep='\\t',names=['FID','IID','pheno'])\n",
    "    phenotype_type='binary' if len(pheno['pheno'].value_counts())<3 else 'continuous'\n",
    "    data_out_assoc_phenotype_path=data_out_assoc_path+phenotype_name+'/'\n",
    "    \n",
    "    for step_idx in range(1,100):\n",
    "        GAT_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.GAT.result.tsv'.format(step_idx),sep='\\t')\n",
    "        plink_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}'.format(step_idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t')\n",
    "        print(phenotype_name,phenotype_type)\n",
    "        \n",
    "    \n",
    "        plink_result_munge=plink_result[plink_result['TEST']=='ADD'].drop(columns='#CHROM').rename(columns={'TEST':'term','ID':'marker_name','BETA':'coef','SE':'std','T_STAT':'Z','OBS_CT':'nobs'})\n",
    "        plink_result_munge['A2']=plink_result_munge.apply(lambda x: x['ALT'] if x['A1']==x['REF'] else x['REF'],axis=1)\n",
    "        plink_result_munge=plink_result_munge.drop(columns=['REF','ALT'])\n",
    "        plink_result_munge['note']='unphased bialleic'  \n",
    "        result_merge=pd.concat([GAT_result,plink_result_munge])[['marker_name','note','term','POS','Z','std','chisq','coef','df','A1','A2','multi_allele','nobs','P']]\n",
    "        \n",
    "        result_merge_sorted=result_merge.sort_values('P')\n",
    "        result_merge_sorted['step']=step_idx\n",
    "        result_merge_sorted['gene']='a'\n",
    "        break\n",
    "\n",
    "    #'data/out_assoc/{}/step_01.plink.PHENO2.glm.linear'.format(i))\n",
    "    \n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_name</th>\n",
       "      <th>note</th>\n",
       "      <th>term</th>\n",
       "      <th>POS</th>\n",
       "      <th>Z</th>\n",
       "      <th>std</th>\n",
       "      <th>chisq</th>\n",
       "      <th>coef</th>\n",
       "      <th>df</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>multi_allele</th>\n",
       "      <th>nobs</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>113652</th>\n",
       "      <td>6:30165273_C/T</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>30165273.0</td>\n",
       "      <td>-18.029400</td>\n",
       "      <td>0.525335</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-9.471480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>1.786990e-72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106155</th>\n",
       "      <td>rs2523942</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>30033065.0</td>\n",
       "      <td>6.014320</td>\n",
       "      <td>2.019910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.148400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59145</td>\n",
       "      <td>1.817020e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94234</th>\n",
       "      <td>6:29925086_T/G</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>29925086.0</td>\n",
       "      <td>5.878330</td>\n",
       "      <td>1.957390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.506200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>4.166410e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305431</th>\n",
       "      <td>6:32537646_C/T</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>32537646.0</td>\n",
       "      <td>5.231500</td>\n",
       "      <td>0.473076</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.474900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>1.687100e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305557</th>\n",
       "      <td>6:32538123_G/A</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>32538123.0</td>\n",
       "      <td>5.172520</td>\n",
       "      <td>0.806447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.171360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>2.317040e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107779</th>\n",
       "      <td>6:30060629_G/C</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>30060629.0</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.475988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>9.999420e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107737</th>\n",
       "      <td>6:30060364_G/A</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>30060364.0</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.475988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>9.999420e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107709</th>\n",
       "      <td>6:30060021_G/A</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>30060021.0</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.475988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>9.999420e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107877</th>\n",
       "      <td>6:30061202_G/A</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>30061202.0</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.475988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>9.999420e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107835</th>\n",
       "      <td>6:30061038_C/A</td>\n",
       "      <td>unphased bialleic</td>\n",
       "      <td>ADD</td>\n",
       "      <td>30061038.0</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.475988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59469</td>\n",
       "      <td>9.999420e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74440 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           marker_name               note term         POS          Z  \\\n",
       "113652  6:30165273_C/T  unphased bialleic  ADD  30165273.0 -18.029400   \n",
       "106155       rs2523942  unphased bialleic  ADD  30033065.0   6.014320   \n",
       "94234   6:29925086_T/G  unphased bialleic  ADD  29925086.0   5.878330   \n",
       "305431  6:32537646_C/T  unphased bialleic  ADD  32537646.0   5.231500   \n",
       "305557  6:32538123_G/A  unphased bialleic  ADD  32538123.0   5.172520   \n",
       "...                ...                ...  ...         ...        ...   \n",
       "107779  6:30060629_G/C  unphased bialleic  ADD  30060629.0   0.000073   \n",
       "107737  6:30060364_G/A  unphased bialleic  ADD  30060364.0   0.000073   \n",
       "107709  6:30060021_G/A  unphased bialleic  ADD  30060021.0   0.000073   \n",
       "107877  6:30061202_G/A  unphased bialleic  ADD  30061202.0   0.000073   \n",
       "107835  6:30061038_C/A  unphased bialleic  ADD  30061038.0   0.000073   \n",
       "\n",
       "             std  chisq       coef  df A1 A2 multi_allele   nobs             P  \n",
       "113652  0.525335    NaN  -9.471480 NaN  T  C          NaN  59469  1.786990e-72  \n",
       "106155  2.019910    NaN  12.148400 NaN  G  T          NaN  59145  1.817020e-09  \n",
       "94234   1.957390    NaN  11.506200 NaN  G  T          NaN  59469  4.166410e-09  \n",
       "305431  0.473076    NaN   2.474900 NaN  C  T          NaN  59469  1.687100e-07  \n",
       "305557  0.806447    NaN   4.171360 NaN  A  G          NaN  59469  2.317040e-07  \n",
       "...          ...    ...        ...  .. .. ..          ...    ...           ...  \n",
       "107779  0.475988    NaN   0.000035 NaN  C  G          NaN  59469  9.999420e-01  \n",
       "107737  0.475988    NaN   0.000035 NaN  A  G          NaN  59469  9.999420e-01  \n",
       "107709  0.475988    NaN   0.000035 NaN  A  G          NaN  59469  9.999420e-01  \n",
       "107877  0.475988    NaN   0.000035 NaN  A  G          NaN  59469  9.999420e-01  \n",
       "107835  0.475988    NaN   0.000035 NaN  A  C          NaN  59469  9.999420e-01  \n",
       "\n",
       "[74440 rows x 14 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_merge_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_bed_path='data/known_genes_chr6.hg18.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALP\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] File b'data/out_assoc/ALP/step_05.omnibus.assoc.munged' does not exist: b'data/out_assoc/ALP/step_05.omnibus.assoc.munged'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-34-36742e5e466a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0midx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnum_independent\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m         \u001b[0momnibus_assoc_result\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata_out_assoc_phenotype_path\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'step_{:02d}.omnibus.assoc.munged'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m         \u001b[0mplink_assoc_result\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata_out_assoc_phenotype_path\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'step_{:02d}.plink.PHENO2.glm.{}.munged'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'logistic'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mphenotype_type\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'binary'\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m'linear'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0;31m#print(plink_assoc_result)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    683\u001b[0m         )\n\u001b[1;32m    684\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    455\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 457\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    893\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    894\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 895\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    896\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"c\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1137\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1915\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"usecols\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1916\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1917\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1918\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1919\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] File b'data/out_assoc/ALP/step_05.omnibus.assoc.munged' does not exist: b'data/out_assoc/ALP/step_05.omnibus.assoc.munged'"
     ]
    }
   ],
   "source": [
    "loci_info_list=[]\n",
    "for phenotype_name in phenotype_list:\n",
    "    #phenotype_name='ALP'\n",
    "    print(phenotype_name)\n",
    "    data_out_assoc_phenotype_path=data_out_assoc_path+phenotype_name+'/'\n",
    "    num_independent=len(glob.glob(data_out_assoc_phenotype_path+'step_*.aa'))\n",
    "    phenotype_data=pd.read_csv(data_out_assoc_phenotype_path+'phenotype.phe',sep='\\t',header=None,names=['FID','IID','pheno'])\n",
    "    phenotype_type= 'continuous' if len(phenotype_data['pheno'].unique())>3 else 'binary'\n",
    "    \n",
    "    for idx in range(1,num_independent+1):\n",
    "        omnibus_assoc_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.assoc.munged'.format(idx),sep='\\t')\n",
    "        plink_assoc_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}.munged'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t').iloc[:,1:]\n",
    "        #print(plink_assoc_result)\n",
    "        plink_cond_f=open(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(5),'r')\n",
    "        plink_cond=[cond.strip() for cond in plink_cond_f.readlines()]\n",
    "        plink_cond_f.close()\n",
    "        plink_covar=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(1),sep='\\t')\n",
    "            \n",
    "        omnibus_top=omnibus_assoc_result.sort_values('P').iloc[0]\n",
    "        plink_top=plink_assoc_result.sort_values('P').iloc[0]\n",
    "            \n",
    "        assoc_type='plink' if plink_top['P']<omnibus_top['P'] else 'omnibus'\n",
    "        \n",
    "        variant=plink_top['ID'] if assoc_type=='plink' else  omnibus_top['Variant']\n",
    "        pos=bim_aa.loc[plink_top['ID']]['pos'] if assoc_type=='plink' else  omnibus_top['Variant']\n",
    "        p_val=plink_top['P'] if assoc_type=='plink' else omnibus_top['P']\n",
    "        \n",
    "        if assoc_type=='omnibus':\n",
    "            gene_name='HLA_'+variant.split('_')[1]\n",
    "        elif assoc_type=='plink':\n",
    "            if variant[:3]=='HLA':\n",
    "                gene_name='HLA_'+variant[4:].split('*')[0]\n",
    "            else:\n",
    "                # if the variant is itself in HLA transcription region\n",
    "                if gene_assign_HLA.iloc[:,1:].loc[variant].sum()==1:\n",
    "                    gene_name='HLA_'+gene_assign_HLA.iloc[:,1:].loc[variant].sort_values(ascending=False).index[0]\n",
    "                elif gene_assign_HLA.iloc[:,1:].loc[variant].sum()>1:\n",
    "                    raise\n",
    "                # if the variant is itself in non-HLA gene coding region\n",
    "                elif gene_assign_nonHLA.iloc[:,1:].loc[variant].sum()==1:\n",
    "                    gene_name='HLA_'+gene_assign_nonHLA.iloc[:,1:].loc[variant].sort_values(ascending=False).index[0]\n",
    "                elif gene_assign_nonHLA.iloc[:,1:].loc[variant].sum()>1:\n",
    "                    raise\n",
    "                else:             \n",
    "                    r2_list=[]\n",
    "                    for idx_bim,(SNP,row) in enumerate(bim_aa.iterrows()):\n",
    "                        r2=pearsonr(plink_aa.get_geno_marker(variant),plink_aa.get_geno_marker(SNP))[0]**2\n",
    "                        r2_list.append(r2)\n",
    "                    r2_df=pd.DataFrame(r2_list,index=bim_aa.index)    \n",
    "                    \n",
    "                    gene_assign_HLA_select=gene_assign_HLA[r2_df[0]>0.5].iloc[:,1:]\n",
    "                    gene_assign_nonHLA_select=gene_assign_nonHLA[r2_df[0]>0.7].iloc[:,1:]\n",
    "                    # if the variant is in strong LD(>0.7) with a variant in HLA transcription region\n",
    "                    if gene_assign_HLA_select.sum(axis=0).sum()==1: \n",
    "                        gene_name=gene_assign_HLA_select.sum(axis=0).sort_values(ascending=False).index[0]\n",
    "                    elif gene_assign_HLA_select.sum(axis=0).sum()>1:\n",
    "                        gene_assign_HLA_select_select_index=gene_assign_HLA_select.sum(axis=1).sort_values(ascending=False).index[gene_assign_HLA_select.sum(axis=1).sort_values(ascending=False)>0]\n",
    "                        gene_assign_HLA_select_select_index_sorted=r2_df.loc[gene_assign_HLA_select_select_index].sort_values(0,ascending=False).index\n",
    "                        if gene_assign_HLA_select.loc[gene_assign_HLA_select_select_index_sorted[0]].sum()==1:\n",
    "                            gene_name=gene_assign_HLA_select.loc[gene_assign_HLA_select_select_index_sorted[0]].sort_values(ascending=False).index[0]\n",
    "                        else:\n",
    "                            raise\n",
    "                    # if the variant is in moderate LD(>0.5) with a variant in non-HLA gene coding region\n",
    "                    elif gene_assign_nonHLA_select.sum(axis=0).sum()==1:\n",
    "                        gene_name=gene_assign_nonHLA_select.sum(axis=0).sort_values(ascending=False).index[0]\n",
    "                    elif gene_assign_nonHLA_select.sum(axis=0).sum()>1:\n",
    "                        gene_assign_nonHLA_select_select_index=gene_assign_nonHLA_select.sum(axis=1).sort_values(ascending=False).index[gene_assign_nonHLA_select.sum(axis=1).sort_values(ascending=False)>0]\n",
    "                        gene_assign_nonHLA_select_select_index_sorted=r2_df.loc[gene_assign_nonHLA_select_select_index].sort_values(0,ascending=False).index\n",
    "                        if gene_assign_nonHLA_select.loc[gene_assign_nonHLA_select_select_index_sorted[0]].sum()==1:\n",
    "                            gene_name=gene_assign_nonHLA_select.loc[gene_assign_nonHLA_select_select_index_sorted[0]].sort_values(ascending=False).index[0]\n",
    "                        else:\n",
    "                            raise\n",
    "                    #in other cases, assign to the nearest gene\n",
    "                    else:\n",
    "                        gene_bed_copy=gene_bed.copy()\n",
    "                        gene_bed_copy['txStartDist']=(gene_bed_copy['txStart']-pos).abs()\n",
    "                        gene_bed_copy['txEndDist']=(gene_bed_copy['txEnd']-pos).abs()\n",
    "                        gene_bed_copy['txDist']=gene_bed_copy[['txStartDist','txEndDist']].min(axis=1)\n",
    "                        gene_name=gene_bed_copy.sort_values('txDist').iloc[0]['name2']     \n",
    "        \n",
    "        \n",
    "        loci_info={'phenotype_name':phenotype_name,\n",
    "                   \n",
    "                   'variant': variant,\n",
    "                   'pos':pos,\n",
    "                   'P':p_val,\n",
    "                   \n",
    "                   'Gene':gene_name,\n",
    "                    'total':(phenotype_data['pheno']!=-9).sum(),\n",
    "                   'case':(phenotype_data['pheno']==2).sum() if phenotype_type=='binary' else (phenotype_data['pheno']!=-9).sum(),\n",
    "                   'control':(phenotype_data['pheno']==1).sum() if phenotype_type=='binary' else 0,\n",
    "                   'A1/A2':((plink_top['A1'],(bim.loc[plink_top['ID']]['a2'] if plink_top['A1']!=bim.loc[plink_top['ID']]['a2'] else bim.loc[plink_top['ID']]['a1']))) if assoc_type=='plink' else ('-','-'),\n",
    "                   'beta(se)': '{}({})'.format(plink_top['BETA'],plink_top['SE']) if assoc_type=='plink' else '-',\n",
    "                   'chisq(df)':'{}({})({})'.format(omnibus_top['chisq'],omnibus_top['deltaDF'],','.join(omnibus_top['Residues'])) if assoc_type=='omnibus' else '-',\n",
    "                   'conditional_covariate':plink_cond+plink_covar.columns.tolist()\n",
    "                  }\n",
    "        \n",
    "        loci_info_list.append(loci_info)\n",
    "        if idx==num_independent:\n",
    "            if p_val>5e-8:\n",
    "                print('error')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
